### ISSUE TITLE
GetDbConnection() and connection pool management

### ISSUE THREAD

**Issue by bragma** [dotnet/efcore#7810]
Hi, I've seen many uses of GetDbConnection() wrapped in a using block. Since disposing the returned connection closes it, I am expecting this is required for some resource management in the connection pool but I've found no explicit explanation of what GetDbConnection does internally (ex: increment a reference counter) and if the "using" is really required in case DI is used for managing the (scoped) lifecycle of the DbContext.

Ex:
In an ASP NET Core app I interweave some use of EF and Dapper and I am in doubt on how to use GetDbConnection(). The pattern I've found on the net is:

```
using (var conn = _dbContext.Database.GetDbConnection())
{
	if (conn.State != ConnectionState.Open)
	{
		await conn.OpenAsync();
	}

	// Dapper query here
}
```

But this causes problems since after the "using" EF does not work anymore (the connection is closed) in the same scope (asp next controller). I was considering if the using can be removed and still not cause any connection leak:

```
var conn = _dbContext.Database.GetDbConnection();
if (conn.State != ConnectionState.Open)
{
	await conn.OpenAsync();
}

// Dapper query here
```

 i.e., can the using be safely removed and still expect connection to be properly handled when the asp net controller method using the DbContext has completed?
Thanks!

**Comment by ajcvickers** [comment]
@bragma If EF creates the DbConnection object, then EF will ensure it is disposed when the DbContext is disposed. On the other hand, if some other code creates the DbConnection object and passes it to EF, then it is the responsibility of the other code to also dispose the connection appropriately.

Similar rules apply to opening and closing the connection. If EF opens the connection, then EF will close the connection when it is done with it. If your code opens the connection, then your code should close the connection.

**Comment by bragma** [comment]
@ajcvickers thanks for the prompt answer. Pardon my ignorance and for resuming a closed question. Given that disposing the DbContext will also dispose the DbConnection (if created by EF) can I assume this will also close the connection, even if I opened it explicitly after getting it with GetDbConnection()?
Thanks again.

**Comment by ajcvickers** [comment]
@bragma Yes, although it would be considered best practice to pair an explicit open and close.

**Comment by bragma** [comment]
Sorry, I have mistakenly deleted my second question. As a reference, I am adding it back:

> Given that DI will dispose DbContext and this in turn will dispose the EF created DbConnection in it, if the DbConnection is obtained via GetDbConnection() can we assume that even if it opened explicitly with conn.open() it will be closed automatically when DbConnection is disposed (so in this case when DbContext will be disposed)?

Answer by @ajcvickers is:

"Yes, although it would be considered best practice to pair an explicit open and close."

Thanks a lot!

**Comment by peteralbanese** [comment]
I was having the same issue as above and changed my code to the following. Hope this helps someone.

public static DataTable GetDataTable (this DbContext context, string sql, Dictionary<string, object> parameters)
        {

            DataTable dt = new DataTable();
            DbConnection cn = context.Database.GetDbConnection();

            using (DbCommand cmd = cn.CreateCommand())
            {
                cmd.CommandText = sql;
                if (parameters != null)
                {
                    foreach (var parameter in parameters)
                    {
                        DbParameter dbParam = cmd.CreateParameter();
                        dbParam.ParameterName = parameter.Key;
                        dbParam.Value = parameter.Value;
                        cmd.Parameters.Add(dbParam);
                    }
                    if (cn.State.Equals(ConnectionState.Closed)) { cn.Open(); }
                    using (DbDataReader reader = cmd.ExecuteReader())
                    {
                        dt.Load(reader);
                    }
                }
            }

            if (cn.State.Equals(ConnectionState.Open)) { cn.Close(); }

            return dt;
        }

**Comment by Rudrik-Andharia** [comment]
I have the same scenario. DbConnection is obtained via GetDbConnection() and it is opened explicitly with conn.open(). But I think DI does not dispose my DbConnection properly.
I'm using .net core 2.2.
In Startup, context is added with default method, so context has 'Scoped' lifetime.
`services.AddDbContext<Context>();`
I think It should be disposed after a request is served.

My connection creates some global temporary tables, Which I can see in System Databases -> tempdb -> TemporaryTables.
After request is served. These tables are still there. Shouldn't they get deleted automatically.

**Comment by ajcvickers** [comment]
@Rudrik-Andharia Please open a new issue and include a small, runnable project/solution or complete code listing that demonstrates the behavior you are seeing.

**Comment by lauxjpn** [comment]
> @bragma If EF creates the DbConnection object, then EF will ensure it is disposed when the DbContext is disposed. On the other hand, if some other code creates the DbConnection object and passes it to EF, then it is the responsibility of the other code to also dispose the connection appropriately.
>
> Similar rules apply to opening and closing the connection. If EF opens the connection, then EF will close the connection when it is done with it. If your code opens the connection, then your code should close the connection.

This should probably make it in the docs (an article about reusing a `DbContext` connection and related stuff might be helpful), or at least in the [RelationalDatabaseFacadeExtensions.GetDbConnection(DatabaseFacade)](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationaldatabasefacadeextensions.getdbconnection?view=efcore-3.1) API reference.

**Comment by kamyker** [comment]
That's a lot of confusing answers. Check it myself to be sure and in the end this isn't true when DbContext is created with dependency injection:
> Similar rules apply to opening and closing the connection. If EF opens the connection, then EF will close the connection when it is done with it. If your code opens the connection, then your code should close the connection.

So the correct version is:
```
var conn = dbContext.Database.GetDbConnection();
if (conn.State != ConnectionState.Open)
{
	await conn.OpenAsync();
}
```

**Comment by ajcvickers** [comment]
@kamyker If you're seeing a case where this is not true:

> If EF opens the connection, then EF will close the connection when it is done with it.

Then that's a bug. Please open a new issue and attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by lauxjpn** [comment]
> That's a lot of confusing answers. Checked it myself to be sure. DbContext will close the connection if EF Core/dependency injection created it. In a way this is false:
> 
> > Similar rules apply to opening and closing the connection. If EF opens the connection, then EF will close the connection when it is done with it. If your code opens the connection, then your code should close the connection.

@kamyker It is not really wrong. The answer that you quoted did not state, that EF Core will not close the connection, but that if you opened it, you _should_ also close it (best practice). EF Core is not going to explicitly _close_ it in this case, once it is done using it, but it will _dispose_ it, when the `DbContext` is being disposed (or reset) and the connection was created earlier by EF Core, which will implicitly close the connection at that point.

(See the [second question and answer](https://github.com/dotnet/efcore/issues/7810#issuecomment-285288280) in this thread.)

**Comment by kamyker** [comment]
Not really, you shouldn't close it if DbContext will be used later on. Especially in more complex logic when DbContext is passed to a method or used by extension method.

@lauxjpn It is wrong, by saying:
> Similar rules apply
and stating before that:
>  it is the responsibility of the other code to also dispose the connection appropriately
then:
> If your code opens the connection, then your code should close the connection

I understand it as "EF Core won't close the connection". I saw the second answer but the first answer is still incorrect or at least misleading.

In the end by lack of direct answer this person also got confused:  https://github.com/dotnet/efcore/issues/7810#issuecomment-384909854
